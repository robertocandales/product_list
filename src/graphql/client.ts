import {
  ApolloClient,
  ApolloLink,
  InMemoryCache,
  createHttpLink,
} from '@apollo/client';

const commerceLink = createHttpLink({
  uri: process.env.REACT_APP_API_URL,

  headers: {
    authorization: localStorage.getItem('Auth-Token')
      ? `Bearer ${localStorage.getItem('Auth-Token')}`
      : '',
  },
});

const afterwareLink = new ApolloLink((operation, forward) => {
  return forward(operation).map((response) => {
    const context = operation.getContext();
    const authHeader = context.response.headers.get('Vendure-Auth-Token');

    if (authHeader) {
      localStorage.setItem('Auth-Token', authHeader);
    }
    return response;
  });
});

export const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: ApolloLink.from([afterwareLink, commerceLink]),
});
