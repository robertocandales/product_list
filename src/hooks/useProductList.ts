import { useMutation, useQuery } from '@apollo/client';
import { ProductsResponse } from '../interfaces/products';
import { useOrder } from '../store/orderProvider';
import { useState } from 'react';
import { GET_PRODUCTS } from '../graphql/queries/getProductQuery';
import { ADD_ITEM_TO_ORDER } from '../graphql/mutation/addOrderMutation';

const useProductList = () => {
  const { addToOrder } = useOrder();
  const [snackbarOpen, setSnackbarOpen] = useState(false);

  const [currentPage, setCurrentPage] = useState(1);
  const itemsPerPage = 6;

  const {
    data = {
      products: {
        totalItems: 0,
        items: [],
      },
    },
    loading,
    error,
  } = useQuery<ProductsResponse>(GET_PRODUCTS, {
    variables: {
      options: { skip: (currentPage - 1) * itemsPerPage, take: itemsPerPage },
    },
  });

  const [addItemToOrder, { loading: MutationLoading }] = useMutation(
    ADD_ITEM_TO_ORDER,
    {
      refetchQueries: [
        {
          query: GET_PRODUCTS,
        },
      ],
      onCompleted: (data) => {
        addToOrder(data?.addItemToOrder);
      },
    }
  );

  const totalPages = Math.ceil(data?.products?.totalItems / itemsPerPage);

  const handleChangePage = (
    event: React.ChangeEvent<unknown>,
    page: number
  ): void => {
    setCurrentPage(page);
  };

  const handleSnackbarClose = () => {
    setSnackbarOpen(false);
  };

  const handleBuyClick = (productId: string) => {
    addItemToOrder({
      variables: { productVariantId: productId, quantity: 1 },
    }).catch((error) => {
      console.log(error, 'error in addItemToOrder');
      setSnackbarOpen(true);
    });
  };
  return {
    productsList: data?.products?.items ?? [],
    loading,
    error,
    snackbarOpen,
    handleBuyClick,
    handleSnackbarClose,
    MutationLoading,
    currentPage,
    itemsPerPage,
    handleChangePage,
    totalPages,
  };
};

export default useProductList;
