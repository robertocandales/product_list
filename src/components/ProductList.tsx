import {
  Grid,
  CardMedia,
  CardContent,
  Typography,
  CardActions,
  Button,
  CircularProgress,
  Alert,
  Snackbar,
  Stack,
  Pagination,
  useMediaQuery,
  useTheme,
} from '@mui/material';

import useProductList from '../hooks/useProductList';
import { formatToUSD } from '../utils/currency';
import { Product } from '../interfaces/products';
import { ProductCard } from './styles';

export function ProductList() {
  const { breakpoints } = useTheme();
  const {
    productsList,
    loading,
    error,
    snackbarOpen,
    handleBuyClick,
    handleSnackbarClose,
    MutationLoading,
    currentPage,
    totalPages,
    handleChangePage,
  } = useProductList();
  const isMobile = useMediaQuery(breakpoints.up('xs'));

  if (loading) {
    return (
      <Grid
        data-testid="loading-spinner"
        container
        justifyContent="center"
        alignItems="center"
        style={{ height: '100vh' }}
      >
        <CircularProgress />
      </Grid>
    );
  }

  if (error) {
    return (
      <Grid
        container
        justifyContent="center"
        alignItems="center"
        style={{ height: '100vh' }}
      >
        <Alert severity="error">
          An error occurred while fetching products: {error.message}
        </Alert>
      </Grid>
    );
  }

  return (
    <Stack alignItems="center">
      <Typography py={[3, 2]} variant={isMobile ? 'h2' : 'h1'}>
        Product List
      </Typography>
      <Grid container spacing={4} sx={{ padding: 4 }}>
        {productsList?.map((product: Product) => (
          <Grid item key={product.id} xs={12} sm={6} md={4} sx={{}}>
            <ProductCard
              sx={{
                height: '100%',
                display: 'flex',
                flexDirection: 'column',
                '&:hover': {
                  boxShadow: '0px 0px 10px rgba(0, 0, 0, 0.2)',
                  transform: 'scale3d(1.05, 1.05, 1)',
                  transition: 'transform 0.3s',
                },
              }}
            >
              <CardMedia
                component="img"
                height="140"
                image={product.featuredAsset.source}
                alt={product.name}
              />
              <CardContent>
                <Typography gutterBottom variant="h5" component="div">
                  {product.name}
                </Typography>
                <Typography variant="body2" color="text.secondary">
                  {product.description}
                </Typography>
                <Stack pt={2}>
                  <Typography variant="body2" color="text.secondary">
                    <strong>Price variants:</strong>
                  </Typography>
                  <Grid container spacing={1} sx={{ marginTop: 1 }}>
                    {product.variants.map((variant, index: number) => (
                      <Grid item key={index} xs={6}>
                        <Typography variant="body2" color="text.secondary">
                          {formatToUSD(variant.price)}
                        </Typography>
                      </Grid>
                    ))}
                  </Grid>
                </Stack>
              </CardContent>
              <CardActions
                sx={{
                  height: '100%',
                  display: 'flex',
                  alignItems: 'flex-end',
                  justifyContent: 'center',
                }}
              >
                <Button
                  variant="contained"
                  data-testid="buy-button-1"
                  onClick={() => handleBuyClick(product.id)}
                  size="small"
                  disabled={MutationLoading}
                >
                  Buy
                </Button>
              </CardActions>
            </ProductCard>
          </Grid>
        ))}
        <Snackbar
          open={snackbarOpen}
          autoHideDuration={6000}
          onClose={handleSnackbarClose}
          message={'An error occurred while adding the item to the order'}
        />
        <Stack py={4} justifyContent="center" alignItems="center" width="100%">
          <Pagination
            count={totalPages}
            page={currentPage}
            onChange={handleChangePage}
          />
        </Stack>
      </Grid>
    </Stack>
  );
}
