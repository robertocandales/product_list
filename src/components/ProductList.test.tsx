import { fireEvent, render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { ProductList } from './ProductList';
import useProductList from '../hooks/useProductList';

jest.mock('../hooks/useProductList', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const mockedUseProductList = useProductList as jest.MockedFunction<
  typeof useProductList
>;

beforeEach(() => {
  mockedUseProductList.mockReturnValue({
    loading: false,
    error: undefined,
    snackbarOpen: false,
    handleBuyClick: jest.fn(),
    handleSnackbarClose: jest.fn(),
    MutationLoading: false,
    productsList: [
      {
        __typename: 'Product',
        id: '1',
        name: 'Product 1',
        description: 'Description 1',
        variants: [
          {
            price: 100,
            __typename: 'ProductVariant',
            productId: '',
          },
        ],
        featuredAsset: {
          source: 'image1.jpg',
          __typename: 'Asset',
          width: 0,
          height: 0,
        },
      },
      {
        __typename: 'Product',
        id: '2',
        name: 'Product 2',
        description: 'Description 2',
        variants: [{ price: 200, __typename: 'ProductVariant', productId: '' }],
        featuredAsset: {
          source: 'image2.jpg',
          __typename: 'Asset',
          width: 0,
          height: 0,
        },
      },
    ],
    currentPage: 1,
    itemsPerPage: 10,
    handleChangePage: jest.fn(),
    totalPages: 1,
  });
});

const Component = () => {
  return (
    <MockedProvider mocks={[]} addTypename={false}>
      <ProductList />
    </MockedProvider>
  );
};

describe('ProductList component', () => {
  it('should render correctly with default props', () => {
    const { container } = render(<Component />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render loading state correctly', () => {
    mockedUseProductList.mockReturnValueOnce({
      loading: true,
      productsList: [],
      error: undefined,
      snackbarOpen: false,
      handleBuyClick: jest.fn(),
      handleSnackbarClose: jest.fn(),
      MutationLoading: false,
      currentPage: 1,
      itemsPerPage: 10,
      handleChangePage: jest.fn(),
      totalPages: 1,
    });
    const { getByTestId } = render(<Component />);
    expect(getByTestId('loading-spinner')).toBeInTheDocument();
  });

  it('should render error state correctly', () => {
    mockedUseProductList.mockReturnValueOnce({
      error: new Error('Test error'),
      currentPage: 1,
      itemsPerPage: 10,
      handleChangePage: jest.fn(),
      totalPages: 1,
    });
    const { getByText } = render(<Component />);
    expect(
      getByText(/An error occurred while fetching products/i)
    ).toBeInTheDocument();
  });

  it('should render products correctly', () => {
    const { getByText } = render(<Component />);
    expect(getByText('Product 1')).toBeInTheDocument();
    expect(getByText('Product 2')).toBeInTheDocument();
  });

  it('should display snackbar on error when adding item to order', () => {
    mockedUseProductList.mockReturnValueOnce({
      snackbarOpen: true,
      productsList: [],
      loading: false,
      error: undefined,
      handleBuyClick: jest.fn(),
      handleSnackbarClose: jest.fn(),
      MutationLoading: false,
      currentPage: 1,
      itemsPerPage: 10,
      handleChangePage: jest.fn(),
      totalPages: 1,
    });
    const { getByText } = render(<Component />);
    expect(
      getByText('An error occurred while adding the item to the order')
    ).toBeInTheDocument();
  });

  it('should call handleBuyClick when Buy button is clicked', () => {
    const handleBuyClick = jest.fn();
    mockedUseProductList.mockReturnValueOnce({
      handleBuyClick,
      productsList: [
        {
          __typename: 'Product',
          id: '1',
          name: 'Product 1',
          description: 'Description 1',
          variants: [
            {
              price: 100,
              __typename: 'ProductVariant',
              productId: '',
            },
          ],
          featuredAsset: {
            source: 'image1.jpg',
            __typename: 'Asset',
            width: 0,
            height: 0,
          },
        },
      ],
      loading: false,
      error: undefined,
      snackbarOpen: false,
      handleSnackbarClose: jest.fn(),
      MutationLoading: false,
      currentPage: 1,
      itemsPerPage: 10,
      handleChangePage: jest.fn(),
      totalPages: 1,
    });
    const { getByTestId } = render(<Component />);
    fireEvent.click(getByTestId('buy-button-1'));
    expect(handleBuyClick).toHaveBeenCalledWith('1');
  });
});
