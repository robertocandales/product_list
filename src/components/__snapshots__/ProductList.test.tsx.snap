// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component should render correctly with default props 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-135gjxb-MuiGrid-root"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-iltpxr-MuiGrid-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-ai43lw-MuiPaper-root-MuiCard-root"
    >
      <img
        alt="Product 1"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="140"
        src="image1.jpg"
      />
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Product 1
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Description 1
        </p>
        <div
          class="css-10yhw90-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            <strong>
              Price variants:
            </strong>
          </p>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1w7ngrt-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                $100.00
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1w45a9n-MuiCardActions-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root css-11qr2p8-MuiButtonBase-root-MuiButton-root"
          data-testid="buy-button-1"
          tabindex="0"
          type="button"
        >
          Buy
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 css-iltpxr-MuiGrid-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-ai43lw-MuiPaper-root-MuiCard-root"
    >
      <img
        alt="Product 2"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="140"
        src="image2.jpg"
      />
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          Product 2
        </div>
        <p
          class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
        >
          Description 2
        </p>
        <div
          class="css-10yhw90-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            <strong>
              Price variants:
            </strong>
          </p>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-1w7ngrt-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-ayy8wd-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
              >
                $200.00
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1w45a9n-MuiCardActions-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root css-11qr2p8-MuiButtonBase-root-MuiButton-root"
          data-testid="buy-button-1"
          tabindex="0"
          type="button"
        >
          Buy
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="css-f6uyne-MuiStack-root"
  >
    <nav
      aria-label="pagination navigation"
      class="MuiPagination-root MuiPagination-text css-jyhyqj-MuiPagination-root"
    >
      <ul
        class="MuiPagination-ul css-wjh20t-MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateBeforeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="page 1"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            tabindex="0"
            type="button"
          >
            1
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateNextIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
